### PR TITLE
Add exporting to latex template

### LINKED ISSUE #66: Adding latex template support
Hi there, 

I use this plugin with templates like Eisvogel https://github.com/Wandmalfarbe/pandoc-latex-template/tree/master 

I was thinking that some of these templates could be integrated as part of the plugin. What are your thoughts? I'm not sure how the user could interact with these in terms of UI. In terms of functionality I think we can modify the pandoc code. 

Alternatively the user could specify their own templates and add them to a folder. 

Let me know what you think

### PR DESCRIPTION
**Pull request by universvm**
Closes #66 .

### CODE REVIEW

File: src/export_command_templates.ts [hunk 1]
```diff
@@ -59,7 +59,6 @@ export default {
     type: 'pandoc',
     arguments:
       '-f markdown --resource-path="${currentDir}" --resource-path="${attachmentFolderPath}" --lua-filter="${luaDir}/pdf.lua" --embed-resources --standalone --metadata title="${currentFileName}" -s -o "${outputPath}" -t pdf',
-    customArguments: '--pdf-engine=typst',
```
Comment (universvm) [inline src/export_command_templates.ts:62]: @mokeyish Why is this in the custom arguments for PDF?

Comment (universvm) [reply src/export_command_templates.ts:62]: But why would `--pdf-engine=typst` be the default?

Comment (mokeyish) [reply src/export_command_templates.ts:62]: You can remove it.

**Comment by mokeyish** [comment]
Thanks for the contribution👍

**Comment by universvm** [comment]
Hi @mokeyish  this is still a draft. Need to do a bit more work

**Comment by universvm** [comment]
Hi @mokeyish ,

I am currently waiting on pandoc to see if I can figure out how to use relative paths: https://github.com/jgm/pandoc

Have tried a lot but don't seem to be able to export. Essentially if you use iCloud, your files are placed in a folder with a space in its name. Latex does not like spaces. 

So far I've built two templates. Here are some sample documents

[paper.pdf](https://github.com/mokeyish/obsidian-enhancing-export/files/11666539/paper.pdf)
[dissertation.pdf](https://github.com/mokeyish/obsidian-enhancing-export/files/11666546/dissertation.pdf)

Generated from this
[paper.md](https://github.com/mokeyish/obsidian-enhancing-export/files/11666551/paper.md)

Regarding your points. I updated the UI to this:
Screenshot 2023-06-06 at 16 42 14

Initially I thought it would be best to have the export templates in export_command_templates as you mentioned. But I realised after showing it to people that it was not intuitive. People seem to see "Type" as the file format and it is not intuitive to scroll through all the extensions to then find the correct pdf. 

So I ended up creating a conditional element in the UI. If the format can take the template, then you see the above image. Else:

Screenshot 2023-06-06 at 16 45 00

Let me know your thoughts. Also I don't think I will be able to work on this much more. Once I find a fix for the template problem I will push it though not sure when.

**Comment by mokeyish** [comment]
- Try to add the path of `.sty` to `--resource-path`, it maybe work.
- You may have misunderstood what I meant. I mean just put the latex template definition into that file.
- `latex template` selector should be condintional emelent, which is why I asked you whether all types support latex templates at the beginning.

**Comment by universvm** [comment]
1. The problem is the space in the file path `/Users/leo/Library/Mobile Documents/` I believe. I did try with resource-path but no luck...
2. - 3. Sorry, I am not sure what you mean. Do you want to define the template path for every export setting?

**Comment by universvm** [comment]
@mokeyish I found a workaround using TEXINPUTS. TEXINPUTS is a variable for setting additional style paths (which is what we need to do here. 

On terminal I modified my .zshrc with this:

```
export TEXINPUTS=/Users/leo/Library/Mobile\ Documents/iCloud~md~obsidian/Documents/exports/.obsidian/plugins/obsidian-enhancing-export/textemplate//:./:$TEXINPUTS
```

Is this something we can help me with? I really need some help here because I don't know JavaScript...

**Comment by mokeyish** [comment]
> @mokeyish I found a workaround using TEXINPUTS. TEXINPUTS is a variable for setting additional style paths (which is what we need to do here.
> 
> On terminal I modified my .zshrc with this:
> 
> ```
> export TEXINPUTS=/Users/leo/Library/Mobile\ Documents/iCloud~md~obsidian/Documents/exports/.obsidian/plugins/obsidian-enhancing-export/textemplate//:./:$TEXINPUTS
> ```
> 
> Ideally we just need to use this variable rather than the path ${textemplateDir}
> 
> Is this something we can help with? I really need some help here because I don't know JavaScript / Apps set this environment variables...


You can set environment variables with following code. (Maybe should make it configurable in the Settings Tab)

```js
await exec('commands', { env: 'TEXINPUTS': 'path/to/textemplate'})
```
https://github.com/mokeyish/obsidian-enhancing-export/blob/8345d2485a8ff0ecede4a2804d6cdf7db1a72ac5/src/utils.ts#L24

**Comment by mokeyish** [comment]
> 1. The problem is the space in the file path `/Users/leo/Library/Mobile Documents/` I believe. I did try with resource-path. It works if the path has no space but iCloud keeps everything in the "Mobile Documents" folder...
> 
>     2. and 3. Sorry, I am not sure what you mean. Do you want to define the template path for every export setting?

It seems that my English is really poor, I can't explain my meaning correctly😂.

**Comment by mokeyish** [comment]
You should pass your `${textemplateDir} ` here just as the work directory(cmd). https://github.com/mokeyish/obsidian-enhancing-export/blob/8345d2485a8ff0ecede4a2804d6cdf7db1a72ac5/src/exporto0o.ts#L171

The concatination of multiple textplateDirs maybe different on different platform(unix: join paths with `:`, windows: join paths with`;`).

**Comment by mokeyish** [comment]
Saw your email, don't worry. Please rewrite following code:

https://github.com/mokeyish/obsidian-enhancing-export/blob/f0ce1d060ccc7acb5e9be71a7965e8a7892941f8/src/exporto0o.ts#L171-L171

as:
```ts
// orginal `$TEXINPUTS` should be empty, so ignore it.
await exec(cmd, { cwd: variables.currentDir, env: { TEXINPUTS: joinEnvPath('path/to/texttemplates/', './') } }); 
```

**Comment by universvm** [comment]
Hey @mokeyish I think we are close! I changed the variable to this:

```
 await exec(cmd, { cwd: variables.currentDir, env: { TEXINPUTS: joinEnvPath(`${textemplateDir}`, `${process.env.TEXINPUTS}`) } }); 

```

Without the previous TEXINPUTS, latex throws errors like `no article.cls` available. 

The problem at the moment is that it says "Success!" and no file is produced. Do you know what could be happening?

File: src/exporto0o.ts [hunk 2]
```diff
@@ -163,9 +163,11 @@ export async function exportToOo(
   }
 
   try {
-    await exec(cmd, { cwd: variables.currentDir, env: { TEXINPUTS: joinEnvPath(`${textemplateDir}`, './') } }); 
+    const { exec } = require('child_process');
```
Comment (mokeyish) [inline src/exporto0o.ts:166]: Remove following lines:
```ts
    const { exec } = require('child_process');
    const path = require('path');
```

We use the `exec` from './utils'

File: src/exporto0o.ts [hunk 3]
```diff
@@ -163,9 +163,11 @@ export async function exportToOo(
...
+    const path = require('path');
     console.log(`[${plugin.manifest.name}]: export command: ${cmd}`);
-    await exec(cmd, { cwd: variables.currentDir });
+    // orginal `$TEXINPUTS` should be empty, so ignore it.
+    await exec(cmd, { cwd: variables.currentDir, env: { TEXINPUTS: joinEnvPath(`${textemplateDir}`, `${process.env.TEXINPUTS}`) } });
```
Comment (mokeyish) [inline src/exporto0o.ts:170]: The
```js
`${textemplateDir}`
``` 
is Javascript [string template](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals), when only one variable, you can use `textemplateDir` directly.

File: src/exporto0o.ts [hunk 4]
```diff
@@ -163,9 +163,11 @@ export async function exportToOo(
   }
 
   try {
-    await exec(cmd, { cwd: variables.currentDir, env: { TEXINPUTS: joinEnvPath(`${textemplateDir}`, './') } }); 
+    const { exec } = require('child_process');
+    const path = require('path');
     console.log(`[${plugin.manifest.name}]: export command: ${cmd}`);
-    await exec(cmd, { cwd: variables.currentDir });
+    // orginal `$TEXINPUTS` should be empty, so ignore it.
```
Comment (mokeyish) [inline src/exporto0o.ts:169]: You can also add `debugger;console.log('current TEXINPUTS', process.env.TEXINPUTS)` at this line. Then open DevTools to see the value of `TEXINPUTS`.

**Comment by mokeyish** [comment]
> Hey @mokeyish I think we are close! I changed the variable to this:
> 
> ```ts
>  await exec(cmd, { cwd: variables.currentDir, env: { TEXINPUTS: joinEnvPath(`${textemplateDir}`, `${process.env.TEXINPUTS}`) } }); 
> ```
> 
> Without the previous TEXINPUTS, latex throws errors like `no article.cls` available.
> 
> The problem at the moment is that it says "Success!" and no file is produced. Do you know what could be happening?


Are you sure it's an environment variable issue? Obsidian's environment variables are only the most basic, and we have to see what is missing and add the rest.


just like

https://github.com/mokeyish/obsidian-enhancing-export/blob/f0ce1d060ccc7acb5e9be71a7965e8a7892941f8/src/main.ts#L32-L39

File: src/exporto0o.ts (hunk 3 above)
Comment (universvm) [reply src/exporto0o.ts]: Thanks - should be fixed now!

**Comment by universvm** [comment]
@mokeyish I think I solved our problems! Can you test it out on your machine to check that things work?

**Comment by mokeyish** [comment]
Thanks for your hard work. There are still some improvements, I will update later.

**Comment by mokeyish** [comment]
@universvm  Hi Leo, Please have a look at the latest code, now you can add Dropdown through configuration.

Here is the full command template.

```ts
{
    name: 'PDF',
    type: 'pandoc',
    arguments:
      '-f markdown --resource-path="${currentDir}" --resource-path="${attachmentFolderPath}" --lua-filter="${luaDir}/pdf.lua" ${ options.textemplate ? `--template="${pluginDir}/textemplate/${options.textemplate}"` : `` } --embed-resources --standalone --metadata title="${currentFileName}" -s -o "${outputPath}" -t pdf',
    customArguments: '--pdf-engine=typst',
    optionsMeta: {
      'textemplate': {
        title: 'Latex Template',
        type: 'dropdown',
        options: [
          { name: 'None', value: null },//
          { name: 'Dissertation', value: 'dissertation.tex' },
          { name: 'Academic Paper', value: 'academic.tex' }
        ]
      }
    },
    extension: '.pdf',
  }

```
There are two points above.

1. Add following snippet to `arguments`
   ```text
   ${ options.textemplate ? `--template="${pluginDir}/textemplate/${options.textemplate}"` : `` }
   ```
2. Add following snippet to optionsMeta

   ```js

   {
     optionsMeta: {
      'textemplate': {
        title: 'Latex Template',
        type: 'dropdown',
        options: [
          { name: 'None', value: null },//
          { name: 'Dissertation', value: 'dissertation.tex' },
          { name: 'Academic Paper', value: 'academic.tex' }
        ]
      }
    }
   }
   ```


https://github.com/mokeyish/obsidian-enhancing-export/blob/e08b6028d5fb0df2e0777b91324aeb68c1ffc20a/tests/renderTemplate.spec.ts#L10-L18

**Comment by mokeyish** [comment]
Now, you don't need to modify the ExportDialog.tsx file.  It seems that Exporto0o.ts also does not need to be modified.

**Comment by universvm** [comment]
Hi @mokeyish , looks good! are we okay merging then?

**Comment by mokeyish** [comment]
> Hi @mokeyish , looks good! are we okay merging then?

No, Your PR has conflicts with the latest code.

**Comment by universvm** [comment]
@mokeyish  did you get rid of the dropdown? 

Now you cannot select the template:
Screenshot 2023-06-09 at 17 21 27


This was before:

Screenshot 2023-06-09 at 17 20 41

**Comment by universvm** [comment]
@mokeyish  have a look now :)

**Comment by mokeyish** [comment]
@universvm I update your code directly.

In your case, it should work, but it's not good enough.

图片


https://github.com/universvm/obsidian-enhancing-export/blob/ad4abf6d5de8ca30fee02613ee7f1f5188ef5a0f/src/export_templates.ts#L60-L74

**Comment by universvm** [comment]
Hi @mokeyish I don't think this will work. We need to append to TEXINPUTS, not replace it.

**Comment by universvm** [comment]
You pushed into my branch and now the code does not work. Please don't merge...

**Comment by universvm** [comment]
@mokeyish  Apologies I misunderstood what you meant.

Your code does not work. As I mentioned. You need to append and not replace TEXINPUTS.

Please keep it as:

```
await exec(cmd, { cwd: variables.currentDir, env: { TEXINPUTS: joinEnvPath(textemplateDir, process.env.TEXINPUTS) } }); 
```

I fixed I think we are ready for merge.

**Comment by mokeyish** [comment]
图片

**Comment by universvm** [comment]
> 图片

?

**Comment by mokeyish** [comment]
图片

What's the value in your machine? Mine is undefined.

**Comment by universvm** [comment]
I know it shows up as undefined but it is not. I've spent like a full day to try to work this out. 

See the pandoc creator: https://github.com/jgm/pandoc/issues/8894#issuecomment-1581139891

And other resources

https://tex.stackexchange.com/questions/93712/definition-of-the-texinputs-variable
https://www2.ph.ed.ac.uk/~wjh/tex/documents/environmental.pdf


Keep it like this:


````
await exec(cmd, { cwd: variables.currentDir, env: { TEXINPUTS: joinEnvPath(textemplateDir, process.env.TEXINPUTS) } }); 

```

**Comment by mokeyish** [comment]
I'm on windows and can't test latex.  Let's do this for now, and optimize it slowly later.

**Comment by universvm** [comment]
Nice, is this good to be merged then?

**Comment by mokeyish** [comment]
> Nice, is this good to be merged then?

It cannot be merged yet, latex templates are currently hard-coded.  We should make it configurable.

**Comment by universvm** [comment]
> > Nice, is this good to be merged then?
> 
> It cannot be merged yet, latex templates are currently hard-coded. We should make it configurable.

There is a dropdown menu. What are you referring to?

Screenshot 2023-06-10 at 08 09 29

File: src/exporto0o.ts [hunk 5]
```diff
@@ -141,23 +129,33 @@ export async function exportToOo(
...
+  const templateOption = templateOptions.find((option) => option.value === exportTemplate); // Use exportTemplate parameter
+  const templatePath = templateOption ? templateOption.path : '';
+  let cmdTpl =
     setting.type === 'pandoc'
-      ? `${pandocPath} ${setting.arguments ?? ''} ${setting.customArguments ?? ''} "${currentPath}"`
+      ? `${pandocPath} ${setting.arguments ?? ''} ${setting.customArguments ?? ''}`
       : setting.command;
-  const cmd = generateCommand(cmdTpl, variables);
+  // if template is selected, append to command
+  if (exportTemplate !== 'none') {
```
Comment (mokeyish) [inline src/exporto0o.ts:148]: This is unnecessary.  The template configuration already does this automatically.


You revert my changes.

File: src/export_templates.ts [hunk 6]
```diff
@@ -58,18 +58,17 @@ export default {
...
-    //   'textemplate': {
-    //     title: 'Latex Template',
-    //     type: 'dropdown',
-    //     options: [
-    //       { name: 'None', value: null },
-    //       { name: 'Dissertation', value: 'dissertation.tex' },
-    //       { name: 'Academic Paper', value: 'academic.tex' }]
-    //   }
-    // },
+      '-f markdown --resource-path="${currentDir}" --resource-path="${attachmentFolderPath}" --lua-filter="${luaDir}/pdf.lua" ${ options.textemplate ? `--resource-path="${pluginDir}/textemplate" --template="${options.textemplate}"` : ` ` } --embed-resources --standalone -s -o "${outputPath}" -t pdf',
```
Comment (mokeyish) [inline src/export_templates.ts:61]: Please look into here. the template will automatically add latex template.

**Comment by universvm** [comment]
Sorry I thought I fixed the conflicts on merge... give me a minute.

File: src/ui/ExportDialog.tsx [hunk 7]
```diff
@@ -10,6 +10,11 @@ import Button from './components/Button';
 import PropertyGrid, { createDefaultObject } from './components/PropertyGrid';
 import Setting, {Text, DropDown, ExtraButton, Toggle} from './components/Setting';
 
+export const templateOptions = [
```
Comment (mokeyish) [inline src/ui/ExportDialog.tsx:13]: Hey man, you shouldn't modified this file. when `optionsMeta` definded a dropdown, the PropertyGrid will automatically render dropdown.

Please double check my leatest code.

**Comment by universvm** [comment]
Ok I can add your changes but as I mentioned before the app does not work. It loads forever when exporting.

**Comment by mokeyish** [comment]
> @universvm Hi Leo, Please have a look at the latest code, now you can add Dropdown through configuration.
> 
> Here is the full command template.
> 
> ```ts
> {
>     name: 'PDF',
>     type: 'pandoc',
>     arguments:
>       '-f markdown --resource-path="${currentDir}" --resource-path="${attachmentFolderPath}" --lua-filter="${luaDir}/pdf.lua" ${ options.textemplate ? `--template="${pluginDir}/textemplate/${options.textemplate}"` : `` } --embed-resources --standalone --metadata title="${currentFileName}" -s -o "${outputPath}" -t pdf',
>     customArguments: '--pdf-engine=typst',
>     optionsMeta: {
>       'textemplate': {
>         title: 'Latex Template',
>         type: 'dropdown',
>         options: [
>           { name: 'None', value: null },//
>           { name: 'Dissertation', value: 'dissertation.tex' },
>           { name: 'Academic Paper', value: 'academic.tex' }
>         ]
>       }
>     },
>     extension: '.pdf',
>   }
> ```
> 
> There are two points above.
> 
> 1. Add following snippet to `arguments`
>    ```
>    ${ options.textemplate ? `--template="${pluginDir}/textemplate/${options.textemplate}"` : `` }
>    ```
> 2. Add following snippet to optionsMeta
>    ```js
>    {
>      optionsMeta: {
>       'textemplate': {
>         title: 'Latex Template',
>         type: 'dropdown',
>         options: [
>           { name: 'None', value: null },//
>           { name: 'Dissertation', value: 'dissertation.tex' },
>           { name: 'Academic Paper', value: 'academic.tex' }
>         ]
>       }
>     }
>    }
>    ```
> 
> https://github.com/mokeyish/obsidian-enhancing-export/blob/e08b6028d5fb0df2e0777b91324aeb68c1ffc20a/tests/renderTemplate.spec.ts#L10-L18

@universvm Hi Leo, 
With the latest code, Adding latex dropdown just change the export template definition above

**Comment by universvm** [comment]
@mokeyish  see the code now. I did what you said but the plugin does not work. Unfortunately I cannot work on this anymore. 

If you can fix the bug feel free to merge. If not close this issue and I will publish my templates somewhere else.

**Comment by mokeyish** [comment]
I also don't really want to spend time on this project 😂.  I wrote this project because I couldn't find a suitable tool for sharing notes and wanted to export it to the hugo blog, so I spent a few days writing such a plugin.


If you don't want to continue to toss.  You can put the final `TEXINPUTS` value and the final export command.  Check it out if I have time.

**Comment by universvm** [comment]
Well then why didn't you merge the commit I made that was working?

**Comment by universvm** [comment]
Well you could have merged the PR when it was working :P

**Comment by universvm** [comment]
If you don't have time, merge this: [92ef758](https://github.com/mokeyish/obsidian-enhancing-export/pull/80/commits/92ef75880b7a52e8f2e030c3c332030bc061374c)

**Comment by universvm** [comment]
@mokeyish  Addressed all the changes you had made + fixed a small bug. 

The templates will be published in here: https://github.com/universvm/academic-pandoc-templates

I will keep this plugin updated with the templates. The repository above contains tutorials on how to use the templates etc..

**Comment by mokeyish** [comment]
Let's close this PR, and could you resubmit a new one base on the branch https://github.com/mokeyish/obsidian-enhancing-export/tree/dev ? 

please use `git rebase` to keep git logs clean. Here is the `git rebase` documentation: https://git-scm.com/docs/git-rebase . Or use git gui such as https://www.syntevo.com/smartgit/, I pretty like smartgit.

**Comment by mokeyish** [comment]
Adding a screenshot, we can see that the commit log without rebase looks very confusing.

图片

File: src/exporto0o.ts [hunk 8]
```diff
@@ -141,23 +126,27 @@ export async function exportToOo(
...
+
   // show progress
   progress.setMessage(lang.preparing(outputFileFullName));
   beforeExport && beforeExport();
   progress.show();
 
   const pandocPath = getPlatformValue(globalSetting.pandocPath) ?? 'pandoc';
 
-  const cmdTpl =
+  let cmdTpl =
```
Comment (mokeyish) [inline src/exporto0o.ts:138]: These changes are unnecessary. The template render function will do this automatically. Do you understand?

Comment (mokeyish) [reply src/exporto0o.ts:138]: The whole changes of this file are also unnecessary.

**Comment by universvm** [comment]
Hi @mokeyish 

Could you have a look now? I reverted the changes you mentioned and tested. 

If all is good I will clean up the commit history. If not let me know I can change

**Comment by universvm** [comment]
@mokeyish I am having problems with conflicts squashing the commits. Can we squash everything into one when merging to dev?

**Comment by mokeyish** [comment]
> @mokeyish I am having problems with conflicts squashing the commits. Can we squash everything into one when merging to dev?

No need to resolve conflicts. create a new branch to applly your changes.

**Comment by universvm** [comment]
> > @mokeyish I am having problems with conflicts squashing the commits. Can we squash everything into one when merging to dev?
> 
> No need to resolve conflicts. create a new branch to applly your changes.

Ah ok! Will probably do these tomorrow :)

**Pull request closed without merging**